File: components/template-store/src/template_store/models/template.py
"""Template model for the Template Store service."""

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any, ClassVar

from beanie import Document
from pydantic import BaseModel, Field, field_validator
from pymongo import IndexModel

from .field import FieldDefinition
from .rule import ValidationRule


class TemplateUsage(StrEnum):
    """How a template's documents are intended to be used.

    - entity (default): full document lifecycle, the v1.x behaviour.
    - reference: lightweight controlled-vocabulary documents (LOV).
      Reserved for a future phase; currently behaves like entity.
    - relationship: typed, property-carrying edge between two
      documents. Requires source_templates / target_templates to be set
      and a source_ref / target_ref reference field on the template.
      See docs/design/document-relationships.md.
    """
    ENTITY = "entity"
    REFERENCE = "reference"
    RELATIONSHIP = "relationship"


class EndpointRef(BaseModel):
    """One declared endpoint of an edge type — a reference with both halves.

    A reference names an identity two ways at once, and both halves are
    load-bearing across a fresh restore:

    - ``lookup_value`` — the caller's anchor, kept verbatim (a template value,
      ``ns:VALUE``, or a canonical id). This is the synonym half: it names the
      identity rather than a specific minting of it, so it survives
      re-anchoring and is the only resolvable handle for endpoints that are
      not inside an archive being restored.
    - ``resolved`` — the canonical template_id this lookup resolved to, filled
      by the platform at write time (never by the caller) and rewritten
      through the id mapping table on a fresh restore. ``None`` only for
      drafts awaiting activation, rows written before this field existed, and
      restored declarations whose endpoint does not exist on the target
      (recorded with a job warning; a null half fails closed wherever an
      exact id is required).

    Documents' ``references[]`` store the same two halves for the same
    reasons; this makes the edge-type declaration follow the platform's one
    reference pattern instead of a single string whose form every consumer
    had to guess.
    """

    lookup_value: str = Field(
        description="The endpoint reference as submitted — value, ns:VALUE, or id; kept verbatim"
    )
    resolved: str | None = Field(
        default=None,
        description="Canonical template_id (server-owned; null when not yet / no longer resolvable)"
    )

    @classmethod
    def coerce(cls, item: "EndpointRef | dict | str") -> "EndpointRef":
        """Accept the legacy single-string form as a lookup-only entry.

        Rows and archives written before the two-half shape hold bare
        strings; they hydrate as ``{lookup_value: s, resolved: None}`` and
        heal at the next write, activation, or restore-door repair.
        """
        if isinstance(item, EndpointRef):
            return item
        if isinstance(item, str):
            return cls(lookup_value=item)
        if isinstance(item, BaseModel):
            return cls.model_validate(item.model_dump())
        return cls.model_validate(item)


class ReportingConfig(BaseModel):
    """Configuration for reporting/analytics sync to PostgreSQL."""

    sync_enabled: bool = Field(
        default=True,
        description="Whether to sync documents of this template to PostgreSQL"
    )
    sync_strategy: str = Field(
        default="latest_only",
        description="Sync strategy: 'latest_only' (upsert) or 'all_versions' (insert all)"
    )
    table_name: str | None = Field(
        default=None,
        description="Custom PostgreSQL table name (auto-generated from value if not set)"
    )
    include_metadata: bool = Field(
        default=True,
        description="Include created_at, created_by, etc. columns"
    )
    flatten_arrays: bool = Field(
        default=True,
        description="Flatten arrays into multiple rows (cross-product)"
    )
    max_array_elements: int = Field(
        default=10,
        ge=1,
        le=100,
        description="Maximum array elements to include when flattening"
    )
    cross_version_view: dict[str, Any] | None = Field(
        default=None,
        description=(
            "Opt-in cross-version entity view over the per-version reporting "
            "tables: {'versions': 'all' | [ints], 'columns': {target: "
            "{'from': source} | {}}}. The identity core is always included; "
            "declared column mappings extend it. Consumed by reporting-sync "
            "(which validates the shape); stored pass-through here."
        )
    )


class TemplateMetadata(BaseModel):
    """Additional metadata for a template."""

    domain: str | None = Field(
        default=None,
        description="Business domain (e.g., 'hr', 'finance', 'healthcare')"
    )
    category: str | None = Field(
        default=None,
        description="Template category (e.g., 'master_data', 'transaction')"
    )
    tags: list[str] = Field(
        default_factory=list,
        description="Tags for categorization and search"
    )
    custom: dict[str, Any] = Field(
        default_factory=dict,
        description="Custom metadata fields"
    )


class Template(Document):
    """
    A schema definition for documents.

    Templates define the structure, validation rules, and constraints
    that documents must conform to. They support inheritance, allowing
    child templates to extend parent templates.

    Examples:
    - Person template: name, birth_date, national_id fields
    - Employee template (extends Person): adds employee_id, department
    - Address template: street, city, postal_code, country
    """

    # Namespace for multi-tenant isolation
    namespace: str = Field(
        ...,
        description="Namespace for data isolation (e.g., wip, dev, seed)"
    )

    # Identity (from Registry)
    template_id: str = Field(
        ...,
        description="Unique ID from Registry (UUID by default)"
    )

    # Human-friendly identifier (mutable)
    value: str = Field(
        ...,
        description="Human-readable value (e.g., 'PERSON'). Must be unique within namespace."
    )

    # Display information
    label: str = Field(
        ...,
        description="Display label (e.g., 'Person Template')"
    )
    description: str | None = Field(
        default=None,
        description="Detailed description of the template's purpose"
    )

    # Versioning
    version: int = Field(
        default=1,
        description="Version number, incremented on updates"
    )

    # Inheritance
    extends: str | None = Field(
        default=None,
        description="Parent template ID for inheritance"
    )
    extends_version: int | None = Field(
        default=None,
        description="Pinned parent version. Required (non-null) whenever the template declares 'extends' — a null there is rejected; the parent is validated against this exact pinned version, never 'latest'. Null only when there is no parent."
    )

    # Identity fields for document upsert
    identity_fields: list[str] = Field(
        default_factory=list,
        description="Fields that form the composite identity key for documents"
    )

    # Peer-projection fields (CASE-343). Names of fields the platform
    # should include when projecting this template's documents in compact
    # "header" contexts: relationships endpoint's `?include=peers`,
    # registry-list summaries, etc. Bare names target `data.<name>`;
    # `metadata.custom.<name>` paths are allowed for app-defined audit
    # fields. Empty list / None → projection falls back to identity_fields
    # at read time (template authors who want richer projection declare
    # explicit header_fields; identity-only templates work zero-config).
    header_fields: list[str] = Field(
        default_factory=list,
        description=(
            "Fields to include in peer/header projections. "
            "Bare names → data.<name>; metadata.custom.<name> paths "
            "allowed. Empty → projection falls back to identity_fields."
        )
    )

    # Declared renames for THIS version relative to the previous one:
    # {new_field: old_field}. A rename declaration means "mechanically the
    # same data under a new key" — it makes the rename losslessly
    # auto-migratable (migrate re-keys before target validation) and
    # matview-mappable, where an undeclared rename is indistinguishable
    # from drop+add. Identity fields can never be renamed (identity is
    # immutable across versions). None/empty on version 1 by definition.
    renames: dict[str, str] | None = Field(
        default=None,
        description=(
            "Field renames relative to the previous version, as "
            "{new_field: old_field}. Enables lossless migration of renamed "
            "fields; identity fields cannot appear."
        )
    )

    # Usage annotation — controls validation, query APIs, and reporting
    # shape. Default 'entity' = v1.x behaviour. 'relationship' enables
    # the document-relationship feature (requires source_templates,
    # target_templates, and source_ref/target_ref reference fields).
    # Immutable after creation.
    usage: TemplateUsage = Field(
        default=TemplateUsage.ENTITY,
        description="Usage class: entity (default), reference, or relationship"
    )

    # Relationship templates only — the single declaration of which
    # templates may sit at each end of an edge, stored as two-half
    # references (see EndpointRef): the submitted anchor verbatim plus the
    # server-resolved canonical id. Rows written before the two-half shape
    # hold bare strings and are coerced to lookup-only entries on read.
    source_templates: list[EndpointRef] = Field(
        default_factory=list,
        description="Declared edge-source endpoints (relationship only)"
    )

    # Relationship templates only — declared edge-target endpoints.
    # See source_templates above.
    target_templates: list[EndpointRef] = Field(
        default_factory=list,
        description="Declared edge-target endpoints (relationship only)"
    )

    @field_validator("source_templates", "target_templates", mode="before")
    @classmethod
    def _coerce_endpoint_entries(cls, v: Any) -> Any:
        # Coercion for stored history, not a write-time guard (write-time
        # rules live at the service's write seam): legacy rows hold bare
        # strings, which hydrate as lookup-only entries.
        if isinstance(v, list):
            return [EndpointRef.coerce(item) for item in v]
        return v

    # Whether updates create new versions (true) or overwrite in place
    # (false). Default true matches v1.x behaviour. Immutable after
    # creation — flipping mid-life would silently reshape an existing
    # template's lifecycle.
    versioned: bool = Field(
        default=True,
        description="True = updates create new versions; False = overwrite in place. Immutable after creation."
    )

    # Schema definition
    fields: list[FieldDefinition] = Field(
        default_factory=list,
        description="Field definitions"
    )

    # Cross-field validation rules
    rules: list[ValidationRule] = Field(
        default_factory=list,
        description="Cross-field validation rules"
    )

    # Metadata
    metadata: TemplateMetadata = Field(
        default_factory=TemplateMetadata,
        description="Additional metadata"
    )

    # Reporting configuration
    reporting: ReportingConfig | None = Field(
        default=None,
        description="Configuration for PostgreSQL reporting sync"
    )

    # Lifecycle
    status: str = Field(
        default="active",
        description="Status: draft, active, inactive"
    )
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC)
    )
    created_by: str | None = Field(
        default=None,
        description="User or system that created this template"
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC)
    )
    updated_by: str | None = Field(
        default=None,
        description="User or system that last updated this template"
    )

    class Settings:
        name = "templates"
        indexes: ClassVar[list[IndexModel]] = [
            # Unique (template_id, version) within namespace — stable ID across versions
            IndexModel([("namespace", 1), ("template_id", 1), ("version", 1)], unique=True, name="ns_template_id_version_unique_idx"),
            # Unique value+version within namespace
            IndexModel([("namespace", 1), ("value", 1), ("version", 1)], unique=True, name="ns_value_version_unique_idx"),
            # Value lookup within namespace
            IndexModel([("namespace", 1), ("value", 1)], name="ns_value_idx"),
            # Status filter within namespace
            IndexModel([("namespace", 1), ("status", 1)], name="ns_status_idx"),
            # Extends lookup within namespace
            IndexModel([("namespace", 1), ("extends", 1)], name="ns_extends_idx"),
            # template_id lookup (non-unique, for finding all versions)
            IndexModel([("template_id", 1)], name="template_id_idx"),
            # Text search (global)
            IndexModel([("label", "text"), ("description", "text")], name="text_search_idx"),
        ]

File: libs/wip-client/src/types/template.ts
import type { PaginatedResponse } from './common.js'

export type FieldType =
  | 'string'
  | 'number'
  | 'integer'
  | 'boolean'
  | 'date'
  | 'datetime'
  | 'term'
  | 'reference'
  | 'file'
  | 'object'
  | 'array'

export type ReferenceType = 'document' | 'term' | 'terminology' | 'template'
export type VersionStrategy = 'latest' | 'pinned'

export type SemanticType =
  | 'email'
  | 'url'
  | 'latitude'
  | 'longitude'
  | 'percentage'
  | 'duration'
  | 'geo_point'

export interface FieldValidation {
  pattern?: string
  min_length?: number
  max_length?: number
  minimum?: number
  maximum?: number
  enum?: unknown[]
}

export interface FileFieldConfig {
  allowed_types: string[]
  max_size_mb: number
  multiple: boolean
  max_files?: number
}

export interface FieldDefinition {
  name: string
  label: string
  type: FieldType
  mandatory: boolean
  default_value?: unknown
  terminology_ref?: string
  template_ref?: string
  /** Pinned version of template_ref. Backend (CASE-493) requires this when template_ref is set. */
  template_ref_version?: number
  reference_type?: ReferenceType
  target_templates?: string[]
  target_terminologies?: string[]
  version_strategy?: VersionStrategy
  file_config?: FileFieldConfig
  array_item_type?: FieldType
  array_terminology_ref?: string
  array_template_ref?: string
  /** Pinned version of array_template_ref. Backend (CASE-493) requires this when array_template_ref is set. */
  array_template_ref_version?: number
  array_file_config?: FileFieldConfig
  validation?: FieldValidation
  semantic_type?: SemanticType
  include_subtypes?: boolean
  full_text_indexed?: boolean
  inherited?: boolean
  inherited_from?: string
  metadata: Record<string, unknown>
}

export type RuleType =
  | 'conditional_required'
  | 'conditional_value'
  | 'mutual_exclusion'
  | 'dependency'
  | 'pattern'
  | 'range'

export type ConditionOperator =
  | 'equals'
  | 'not_equals'
  | 'in'
  | 'not_in'
  | 'exists'
  | 'not_exists'

export interface Condition {
  field: string
  operator: ConditionOperator
  value?: unknown
}

export interface ValidationRule {
  type: RuleType
  description?: string
  conditions: Condition[]
  target_field?: string
  target_fields?: string[]
  required?: boolean
  allowed_values?: unknown[]
  pattern?: string
  minimum?: number
  maximum?: number
  error_message?: string
}

export type SyncStrategy = 'latest_only' | 'all_versions'

/**
 * How a template's documents are intended to be used.
 *
 * - `entity` (default): full document lifecycle, the v1.x behaviour.
 * - `reference`: lightweight controlled-vocabulary documents (LOV).
 *   Reserved for a future phase; currently behaves like entity.
 * - `relationship`: typed, property-carrying edge between two
 *   documents (a.k.a. "edge type"). Requires source_templates /
 *   target_templates to be set on the template, plus source_ref /
 *   target_ref reference fields. Immutable after creation.
 *
 * See PoNIF #7 (edge types are stored as templates) and PoNIF #8
 * (`versioned: false` is an option on relationship templates).
 */
export type TemplateUsage = 'entity' | 'reference' | 'relationship'

/**
 * Opt-in cross-version entity view over a template's per-version reporting
 * tables — the config behind the bare `doc_<value>` name.
 *
 * The identity core (the typed intersection of the selected versions'
 * tables) is always included; `columns` adds mappings beyond it. Anything
 * unmapped and not provably identical across the selected versions is
 * absent from the view — the backend never silently merges columns it
 * cannot prove compatible.
 */
export interface CrossVersionView {
  /** Version tables the view spans. Server default is 'all'. */
  versions: 'all' | number[]
  /**
   * Target column → optional source. `{ from: old }` maps a renamed column
   * (declared renames land here). `{}` or `null` means the column keeps its
   * own name in the versions that have it, and is NULL elsewhere.
   */
  columns: Record<string, { from?: string } | null>
}

export interface ReportingConfig {
  sync_enabled: boolean
  sync_strategy: SyncStrategy
  table_name?: string
  include_metadata: boolean
  flatten_arrays: boolean
  max_array_elements: number
  /**
   * Opt-in cross-version view config. Stored pass-through on the template;
   * the shape is owned and validated by reporting-sync, which builds the
   * view. Absent / null means the bare name exposes the identity core only.
   */
  cross_version_view?: CrossVersionView | null
}

export interface TemplateMetadata {
  domain?: string
  category?: string
  tags: string[]
  custom: Record<string, unknown>
}

export interface Template {
  template_id: string
  namespace: string
  value: string
  label: string
  description?: string
  version: number
  extends?: string
  extends_version?: number
  identity_fields: string[]
  /**
   * Fields to surface in peer/header projection contexts (CASE-343).
   * Bare names target `data.<name>`; `metadata.custom.<name>` paths
   * are allowed for audit fields. Empty → the platform's projection
   * falls back to `identity_fields`. The relationships endpoint's
   * `?include=peers` projection reads this; future list / summary
   * endpoints may consume it too.
   */
  header_fields?: string[]
  /**
   * Usage class: entity (default), reference, or relationship.
   * Immutable after creation. Relationship templates ("edge types")
   * additionally require source_templates + target_templates and
   * source_ref / target_ref reference fields.
   */
  usage?: TemplateUsage
  /**
   * Template values allowed as the source endpoint of an edge.
   * Set only on relationship templates; empty / absent on entity and
   * reference templates.
   */
  source_templates?: string[]
  /**
   * Template values allowed as the target endpoint of an edge.
   * Set only on relationship templates.
   */
  target_templates?: string[]
  /**
   * True (default) = updates create new versions; false = overwrite
   * in place. Currently only available on relationship templates.
   * Immutable after creation. See PoNIF #8.
   */
  versioned?: boolean
  /**
   * Field renames this version declared relative to the previous one, as
   * `{new_field: old_field}`. Persisted, so it comes back on read — a
   * template editor renders the declaration it was created with.
   */
  renames?: Record<string, string> | null
  fields: FieldDefinition[]
  rules: ValidationRule[]
  metadata: TemplateMetadata
  reporting?: ReportingConfig
  status: 'draft' | 'active' | 'inactive'
  created_at: string
  created_by?: string
  updated_at: string
  updated_by?: string
}

export interface CreateTemplateRequest {
  value: string
  label: string
  description?: string
  template_id?: string
  version?: number
  namespace: string
  extends?: string
  extends_version?: number
  identity_fields?: string[]
  /**
   * Peer/header-projection fields (CASE-343). Bare names target data.*,
   * `metadata.custom.<name>` paths allowed. Empty → projection falls
   * back to identity_fields.
   */
  header_fields?: string[]
  /** Usage class — defaults to 'entity' on the server when omitted. */
  usage?: TemplateUsage
  /** Required when usage='relationship'; ignored otherwise. */
  source_templates?: string[]
  /** Required when usage='relationship'; ignored otherwise. */
  target_templates?: string[]
  /** Defaults to true. Immutable after creation. See PoNIF #8. */
  versioned?: boolean
  /**
   * Field renames relative to the previous version, `{new_field: old_field}`.
   * Validated against the version being renamed from (the old name existed
   * and is gone, the new one is new, types match, identity fields excluded)
   * and rejected on a first version. A declared rename migrates losslessly
   * and maps in reporting; an undeclared one strands the old column's data.
   */
  renames?: Record<string, string>
  fields?: FieldDefinition[]
  rules?: ValidationRule[]
  metadata?: Partial<TemplateMetadata>
  reporting?: Partial<ReportingConfig>
  created_by?: string
  validate_references?: boolean
  status?: string
}

export interface UpdateTemplateRequest {
  value?: string
  label?: string
  description?: string
  extends?: string
  extends_version?: number
  identity_fields?: string[]
  /** Update peer/header-projection fields (CASE-343). */
  header_fields?: string[]
  /**
   * Field renames the new version declares relative to the current one,
   * `{new_field: old_field}`. This is the path an interactive editor takes —
   * renaming a field on an existing template — so it matters here as much as
   * on create. Same validation as the create path.
   */
  renames?: Record<string, string>
  fields?: FieldDefinition[]
  rules?: ValidationRule[]
  metadata?: Partial<TemplateMetadata>
  reporting?: Partial<ReportingConfig>
  updated_by?: string
}

export type TemplateListResponse = PaginatedResponse<Template>

export interface ValidateTemplateRequest {
  check_terminologies?: boolean
  check_templates?: boolean
}

export interface ValidateTemplateResponse {
  valid: boolean
  template_id: string
  errors: Array<{ field: string; code: string; message: string }>
  warnings: Array<{ field: string; code: string; message: string }>
  will_also_activate?: string[]
}

export interface TemplateUpdateResponse {
  template_id: string
  value: string
  version: number
  is_new_version: boolean
  previous_version?: number
}

export interface ActivationDetail {
  template_id: string
  value: string
  status: string
}

export interface ActivateTemplateResponse {
  activated: string[]
  activation_details: ActivationDetail[]
  total_activated: number
  errors: Array<{ field: string; code: string; message: string }>
  warnings: Array<{ field: string; code: string; message: string }>
}

export interface CascadeResult {
  value: string
  old_template_id: string
  new_template_id?: string
  new_version?: number
  status: string
  error?: string
}

export interface CascadeResponse {
  parent_template_id: string
  parent_value: string
  parent_version: number
  total: number
  updated: number
  unchanged: number
  failed: number
  results: CascadeResult[]
}

/**
 * Live-document impact of a version event, computed advisory-side from
 * document-store.
 *
 * `status: 'unavailable'` is a real answer, not an error: document-store
 * could not be reached, so the counts are unknown. It is deliberately never
 * a silent zero, which would read as "no documents affected".
 */
export interface TemplateVersionImpact {
  status: 'ok' | 'unavailable'
  /** Present when status === 'ok'. */
  total_live_docs?: number
  /** Live document count keyed by the template version they validated against. */
  docs_per_version?: Record<string, number>
  /** Per-field count of live documents where the field is non-empty. */
  field_nonempty_counts?: Record<string, number>
  /** Present when status === 'unavailable' — why the counts are missing. */
  reason?: string
}

/**
 * Whether the platform can offer to move existing documents onto the new
 * version, and how.
 */
export interface TemplateMigrationOffer {
  /**
   * true = offerable; false = needs app-side data decisions (type changes,
   * newly-required fields, removed fields carrying live data); null =
   * removed fields present but the counts were unavailable, so run the
   * dry-run for a per-document readiness report.
   */
  eligible: boolean | null
  reason: string
  /** The operation to run — the migrate dry-run/apply cycle. */
  via: string
}

/**
 * `details` on a bulk result item that minted a new template version: the
 * schema diff, plus the consequences of having minted it.
 *
 * Both write paths carry this — create-as-upsert (`POST /templates`) and
 * update (`PUT /templates`). Older backends attached it on create only, so
 * treat it as optional and narrow with `asVersionEventDetails`.
 */
export interface TemplateVersionEventDetails {
  added_optional: string[]
  added_required: string[]
  removed: string[]
  changed_type: Array<{ name: string; old_type: string; new_type: string }>
  made_required: string[]
  modified_existing: string[]
  identity_changed: { old: string[]; new: string[] } | null
  relationship_refs_changed: unknown | null
  impact: TemplateVersionImpact
  migration: TemplateMigrationOffer
}

/**
 * Narrow a bulk result item's untyped `details` to the version-event shape.
 *
 * `BulkResultItem.details` is `Record<string, unknown>` because the envelope
 * is shared by every bulk endpoint — documents, terms, templates. Rather
 * than widening it with template-specific members, narrow here at the point
 * of use.
 *
 * Returns null when the item did not mint a version, or when the backend
 * predates the impact block on the path that was used.
 */
export function asVersionEventDetails(
  details: Record<string, unknown> | undefined | null,
): TemplateVersionEventDetails | null {
  if (!details) return null
  const impact = details.impact as TemplateVersionImpact | undefined
  const migration = details.migration as TemplateMigrationOffer | undefined
  if (!impact || !migration) return null
  return details as unknown as TemplateVersionEventDetails
}
